cuda/btx_cudainterval_callbacks.cpp

```diff
@@ -0,0 +1,589 @@
+#include <cassert>
+#include <cstdlib>
+#include <iostream>
+#include <regex>
+#include <tuple>
+#include <string>
+#include <unordered_map>
+
+#include "btx_cudainterval_callbacks.hpp"
+
+#include "context_manager.hpp"
+
+#include <metababel/metababel.h>
+
+
+struct data_s {
+  // store entry state per thread in order to use in exit callback, conditioned
+  // on the cuResult being a non-error
+  std::unordered_map<hpt_t, int64_t> entry_ts;
+  std::unordered_map<hpt_t, std::string> entry_kernel_name;
+  std::unordered_map<hpt_t, size_t> entry_traffic_size;
+
+  // kernels (CUfunction objects) are "created" before use, at which time we save
+  // the name so at execute time we know the name and can emit an event with the name
+  std::unordered_map<hp_kernel_t, std::string> hp_kernel_to_name;
+
+  // At device operation launch time, save the device the operation was launched on,
```

Comment:
It's not a "device operation", everything here is running on the host.
Some maybe "Used to save the last device in with the particular function have been launched".

And I guess "function" is bad, we should have use "command". More meaning full I think